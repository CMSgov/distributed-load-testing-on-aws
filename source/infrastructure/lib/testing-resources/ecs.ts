// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, CfnResource, Tags, Fn } from "aws-cdk-lib";
import { ContainerImage, FargateTaskDefinition, LogDriver, CfnCluster } from "aws-cdk-lib/aws-ecs";
import { Effect, ServicePrincipal, Role, PolicyDocument, PolicyStatement } from "aws-cdk-lib/aws-iam";
import { LogGroup, RetentionDays } from "aws-cdk-lib/aws-logs";
import { CfnSecurityGroup, CfnSecurityGroupEgress, CfnSecurityGroupIngress } from "aws-cdk-lib/aws-ec2";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";
import { DockerImageAsset } from "aws-cdk-lib/aws-ecr-assets";
import * as path from "path";
import { addCfnGuardSuppression } from "../common-resources/add-cfn-guard-suppression";

export interface ECSResourcesConstructProps {
  readonly fargateVpcId: string;
  readonly scenariosS3Bucket: string;
  readonly securityGroupEgress: string;
  readonly solutionId: string;
  readonly stableTagCondition: string;
}

/**
 * Distributed Load Testing on AWS Fargate and ECS test runner construct.
 * This creates the ECS cluster, Fargate task definition, and Security Group
 */
export class ECSResourcesConstruct extends Construct {
  public taskClusterName: string;
  public ecsCloudWatchLogGroup: LogGroup;
  public taskDefinitionArn: string;
  public taskExecutionRoleArn: string;
  public ecsSecurityGroupId: string;
  public taskDefinitionFamily: string;

  constructor(scope: Construct, id: string, props: ECSResourcesConstructProps) {
    super(scope, id);

    const dltTaskCluster = new CfnCluster(this, "DLTEcsCluster", {
      clusterName: Aws.STACK_NAME,
      clusterSettings: [{ name: "containerInsights", value: "enabled" }],
      tags: [
        {
          key: "SolutionId",
          value: props.solutionId,
        },
        {
          key: "CloudFormation Stack",
          value: Aws.STACK_NAME,
        },
      ],
    });

    this.taskClusterName = dltTaskCluster.ref;

    const scenariosBucketArn = Bucket.fromBucketName(this, "ScenariosBucket", props.scenariosS3Bucket).bucketArn;

    const dltTaskExecutionRole = new Role(this, "DLTTaskExecutionRole", {
      assumedBy: new ServicePrincipal("ecs-tasks.amazonaws.com"),
      inlinePolicies: {
        ECSTaskExecutionPolicy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                "ecr:GetAuthorizationToken",
                "ecr:BatchCheckLayerAvailability",
                "ecr:GetDownloadUrlForLayer",
                "ecr:BatchGetImage",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              resources: ["*"],
            }),
          ],
        }),
        ScenariosS3Policy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["s3:HeadObject", "s3:PutObject", "s3:GetObject", "s3:ListBucket"],
              resources: [scenariosBucketArn, `${scenariosBucketArn}/*`],
            }),
          ],
        }),
      },
    });
    this.taskExecutionRoleArn = dltTaskExecutionRole.roleArn;

    addCfnGuardSuppression(dltTaskExecutionRole, "IAM_POLICYDOCUMENT_NO_WILDCARD_RESOURCE");
    addCfnGuardSuppression(dltTaskExecutionRole, "IAM_NO_INLINE_POLICY_CHECK");

    this.ecsCloudWatchLogGroup = new LogGroup(this, "DLTCloudWatchLogsGroup", {
      retention: RetentionDays.ONE_YEAR,
    });
    const dltLogsGroupResource = this.ecsCloudWatchLogGroup.node.defaultChild as CfnResource;
    dltLogsGroupResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W84",
          reason: "KMS encryption unnecessary for log group",
        },
      ],
    });
    Tags.of(this.ecsCloudWatchLogGroup).add("SolutionId", props.solutionId);

    const dltTaskDefinition = new FargateTaskDefinition(this, "DLTTaskDefinition", {
      cpu: 2048,
      memoryLimitMiB: 4096,
      executionRole: dltTaskExecutionRole,
      taskRole: dltTaskExecutionRole,
    });
    const dockerRepoName = "distributed-load-testing-on-aws-load-tester";

    const versionTagForImage: string =
      process.env.PUBLIC_ECR_REGISTRY && process.env.PUBLIC_ECR_TAG
        ? `${process.env.PUBLIC_ECR_REGISTRY}/${dockerRepoName}:${process.env.PUBLIC_ECR_TAG}`
        : "";

    const stageTagForImage: string =
      process.env.PUBLIC_ECR_REGISTRY && process.env.PUBLIC_ECR_TAG
        ? `${process.env.PUBLIC_ECR_REGISTRY}/${dockerRepoName}:${
            process.env.PUBLIC_ECR_TAG.substring(0, 4) + "_stable"
          }`
        : "";

    const imageUseTag = Fn.conditionIf(props.stableTagCondition, stageTagForImage, versionTagForImage).toString();

    const imageAsset = imageUseTag
      ? ContainerImage.fromRegistry(`${imageUseTag}`)
      : new DockerImageAsset(this, "LoadTesterImage", {
          directory: path.join(__dirname, `../../../../deployment/ecr/${dockerRepoName}`),
        });

    dltTaskDefinition.addContainer("LoadTestContainer", {
      containerName: `${Aws.STACK_NAME}-load-tester`,
      image: imageAsset instanceof DockerImageAsset ? ContainerImage.fromDockerImageAsset(imageAsset) : imageAsset,
      memoryLimitMiB: 4096,
      logging: LogDriver.awsLogs({
        streamPrefix: "load-testing",
        logGroup: this.ecsCloudWatchLogGroup,
      }),
    });
    Tags.of(dltTaskDefinition).add("SolutionId", props.solutionId);
    this.taskDefinitionArn = dltTaskDefinition.taskDefinitionArn;

    const ecsSecurityGroup = new CfnSecurityGroup(this, "DLTEcsSecurityGroup", {
      vpcId: props.fargateVpcId,
      groupDescription: "DLTS Tasks Security Group",
    });
    ecsSecurityGroup.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W40",
          reason: "IpProtocol set to -1 (any) as ports are not known prior to running tests",
        },
        {
          id: "F1000",
          reason: "egress rule is specificed as its own cfn resource.",
        },
      ],
    });

    this.ecsSecurityGroupId = ecsSecurityGroup.ref;

    new CfnSecurityGroupEgress(this, "DLTSecGroupEgress", {
      cidrIp: props.securityGroupEgress,
      description: "Allow tasks to call out to external resources",
      groupId: ecsSecurityGroup.ref,
      ipProtocol: "-1",
    });

    new CfnSecurityGroupIngress(this, "DLTSecGroupIngress", {
      description: "Allow tasks to communicate",
      fromPort: 50000,
      groupId: ecsSecurityGroup.ref,
      ipProtocol: "tcp",
      sourceSecurityGroupId: ecsSecurityGroup.ref,
      toPort: 50000,
    });

    this.taskDefinitionFamily = dltTaskDefinition.family;
  }
}
