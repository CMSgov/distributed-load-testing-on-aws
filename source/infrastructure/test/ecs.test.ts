// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from 'aws-cdk-lib';
import { Policy, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { ECSResourcesConstruct } from '../lib/testing-resources/ecs';

test('DLT ECS Test', () => {
  const stack = new Stack();
  const testPolicy = new Policy(stack, 'TestPolicy', {
    statements: [
      new PolicyStatement({
        resources: ['*'],
        actions: ['cloudwatch:Get*']
      })
    ]
  });
  const ecs = new ECSResourcesConstruct(stack, 'TestECS', {
    cloudWatchLogsPolicy: testPolicy,
    containerImage: 'testRepository/testImage:testTag',
    fargateVpcId: 'vpc-1a2b3c4d5e',
    scenariosS3Bucket: 'testscenariobucket',
    securityGroupEgress: '0.0.0.0/0',
    solutionId: 'SO0062'
  });
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  expect(stack).toHaveResource('AWS::ECS::Cluster', {
    ClusterSettings: [
      {
        'Name': 'containerInsights',
        'Value': 'enabled'
      }
    ]
  });
  expect(stack).toHaveResource('AWS::IAM::Role', {
    "AssumeRolePolicyDocument": {
      "Statement": [
        {
          "Action": "sts:AssumeRole",
          "Effect": "Allow",
          "Principal": {
            "Service": "ecs-tasks.amazonaws.com"
          }
        }
      ],
      "Version": "2012-10-17"
    }
  });
  expect(stack).toHaveResource('AWS::Logs::LogGroup', {
    "RetentionInDays": 365
  });
  expect(stack).toHaveResource('AWS::ECS::TaskDefinition');
  expect(stack).toHaveResource('AWS::EC2::SecurityGroup');
  expect(stack).toHaveResource('AWS::EC2::SecurityGroupIngress', {
    GroupId: {
      Ref: "TestECSDLTEcsSecurityGroupFE5016DC",
    },
    SourceSecurityGroupId: {
      Ref: "TestECSDLTEcsSecurityGroupFE5016DC",
    }
  });
  expect(stack).toHaveResource('AWS::EC2::SecurityGroupEgress', {
    Description: 'Allow tasks to call out to external resources'
  });
  expect(ecs.taskClusterName).toBeDefined();
  expect(ecs.ecsCloudWatchLogGroup).toBeDefined();
  expect(ecs.taskDefinitionArn).toBeDefined();
  expect(ecs.taskExecutionRoleArn).toBeDefined();
  expect(ecs.ecsSecurityGroupId).toBeDefined();
});