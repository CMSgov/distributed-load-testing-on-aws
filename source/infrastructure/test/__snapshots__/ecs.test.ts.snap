// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DLT ECS Test 1`] = `
Object {
  "Resources": Object {
    "TestECSDLTCloudWatchLogsGroup505959A6": Object {
      "DeletionPolicy": "Retain",
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W84",
              "reason": "KMS encryption unnecessary for log group",
            },
          ],
        },
      },
      "Properties": Object {
        "RetentionInDays": 365,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "TestECSDLTEcsCluster7C6F0F5D": Object {
      "Properties": Object {
        "ClusterName": Object {
          "Ref": "AWS::StackName",
        },
        "ClusterSettings": Array [
          Object {
            "Name": "containerInsights",
            "Value": "enabled",
          },
        ],
        "Tags": Array [
          Object {
            "Key": "CloudFormation Stack",
            "Value": Object {
              "Ref": "AWS::StackName",
            },
          },
          Object {
            "Key": "SolutionId",
            "Value": "SO0062",
          },
        ],
      },
      "Type": "AWS::ECS::Cluster",
    },
    "TestECSDLTEcsSecurityGroupFE5016DC": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W40",
              "reason": "IpProtocol set to -1 (any) as ports are not known prior to running tests",
            },
          ],
        },
      },
      "Properties": Object {
        "GroupDescription": "DLTS Tasks Security Group",
        "VpcId": "vpc-1a2b3c4d5e",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "TestECSDLTSecGroupEgressFC1E4757": Object {
      "Properties": Object {
        "CidrIp": "0.0.0.0/0",
        "Description": "Allow tasks to call out to external resources",
        "GroupId": Object {
          "Ref": "TestECSDLTEcsSecurityGroupFE5016DC",
        },
        "IpProtocol": "-1",
      },
      "Type": "AWS::EC2::SecurityGroupEgress",
    },
    "TestECSDLTSecGroupIngress78B83F3F": Object {
      "Properties": Object {
        "Description": "Allow tasks to communicate",
        "FromPort": 50000,
        "GroupId": Object {
          "Ref": "TestECSDLTEcsSecurityGroupFE5016DC",
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": Object {
          "Ref": "TestECSDLTEcsSecurityGroupFE5016DC",
        },
        "ToPort": 50000,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "TestECSDLTTaskDefinitionB261038E": Object {
      "Properties": Object {
        "ContainerDefinitions": Array [
          Object {
            "Essential": true,
            "Image": "testRepository/testImage:testTag",
            "LogConfiguration": Object {
              "LogDriver": "awslogs",
              "Options": Object {
                "awslogs-group": Object {
                  "Ref": "TestECSDLTCloudWatchLogsGroup505959A6",
                },
                "awslogs-region": Object {
                  "Ref": "AWS::Region",
                },
                "awslogs-stream-prefix": "load-testing",
              },
            },
            "Memory": 4096,
            "Name": Object {
              "Fn::Join": Array [
                "",
                Array [
                  Object {
                    "Ref": "AWS::StackName",
                  },
                  "-load-tester",
                ],
              ],
            },
          },
        ],
        "Cpu": "2048",
        "ExecutionRoleArn": Object {
          "Fn::GetAtt": Array [
            "TestECSDLTTaskExecutionRole104D057B",
            "Arn",
          ],
        },
        "Memory": "4096",
        "NetworkMode": "awsvpc",
        "RequiresCompatibilities": Array [
          "FARGATE",
        ],
        "TaskRoleArn": Object {
          "Fn::GetAtt": Array [
            "TestECSDLTTaskExecutionRole104D057B",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ECS::TaskDefinition",
    },
    "TestECSDLTTaskExecutionRole104D057B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy",
              ],
            ],
          },
        ],
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "s3:HeadObject",
                    "s3:PutObject",
                    "s3:GetObject",
                    "s3:ListBucket",
                  ],
                  "Effect": "Allow",
                  "Resource": Array [
                    Object {
                      "Fn::Join": Array [
                        "",
                        Array [
                          "arn:",
                          Object {
                            "Ref": "AWS::Partition",
                          },
                          ":s3:::testscenariobucket",
                        ],
                      ],
                    },
                    Object {
                      "Fn::Join": Array [
                        "",
                        Array [
                          "arn:",
                          Object {
                            "Ref": "AWS::Partition",
                          },
                          ":s3:::testscenariobucket/*",
                        ],
                      ],
                    },
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "ScenariosS3Policy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestPolicyCC05E598": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "cloudwatch:Get*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestPolicyCC05E598",
        "Roles": Array [
          Object {
            "Ref": "TestECSDLTTaskExecutionRole104D057B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
